### ISSUE TITLE
Possible Biome and Populator API in the future?

### ISSUE THREAD

**Issue by TheShermanTanker** [pmmp/PocketMine-MP#4161]
## Description
Tl;Dr Currently querying about the possibility of exposing the static private function Biome::register to plugins in the future, as well as transferring the Biomes stored in the main thread in Biome::$biomes to the Biome class in the generation thread when GeneratorRegisterTask.php calls Biome::init() (I did try to implement this myself as a PR but Composer is a d*ck and didn't even install the dependencies properly ever after several tries unfortunately)

## Justification
As of now the registering of all known Biomes is hardcoded inside the function Biome::init(), which takes place on server start in the constructor/other initialization function inside the Server class (I forgot exactly which one but you get the point :P). This by itself isn't actually a problem, but the main issue is that due to the level generation system being multithreaded, the generation of levels takes place on a specialized worker thread, with a _second_ separate call to Biome::init() inside GeneratorRegisterTask.php, which means there is a second copy of all biomes and their popolators in that other thread, while the biomes stored on the main thread are entirely unused (To be honest I'm not even sure what they're used for on the main thread). This can lead to confusing instances where generation plugins register their own populators and biomes through Biome::addPopulator() only to have it fail to work silently at runtime without notifying the developer of the true reason as to why. The only 2 workarounds I can recall so far would be to extend Normal.php and overriding several functions  in it, or attempting to access the low level threading API and searching for the static _$biomes_ variable there, the former being very time consuming and the latter being highly impractical for many plugins, so I believe a slight overhaul which allows the Biome class on the main thread to transfer all the biome objects stored to the generation thread would help plugins greatly

## Alternative methods
I did implement this in a very ugly manner quite some time ago, but they relied on passing Biome::$biomes as an unserialized variable to the generation thread constructor, which caused very erratic issues not suitable for production

**Comment by dktapps** [comment]
This issue is talking about multiple problems at the same time.

Right now, it's not possible to register biomes without reflection hacks, which are in no way endorsed or supported. One of the main reasons why there's no API to register biomes is exactly because of the threading problems involved.

>  a slight overhaul which allows the Biome class on the main thread to transfer all the biome objects stored to the generation thread would help plugins greatly

This is not practical. By default, nothing is shared between threads unless shared explicitly (by passing data). This means that you yourself have to either copy biomes from the main thread, or make a callback that fires on each thread.

I think the most suitable solution for this is to use `AsyncPool->addWorkerStartHook()`, which will allow you to schedule an async task to register biomes on that worker before any generation takes place.

**Comment by TheShermanTanker** [comment]
> I think the most suitable solution for this is to use AsyncPool->addWorkerStartHook(), which will allow you to schedule an async task to register biomes on that worker before any generation takes place.

Interesting, I'll see if I can do anything with this in a PR (If composer is willing to work with me, heh)

**Comment by TheShermanTanker** [comment]
About this, I recently tried adding an anonymous function to the AsyncPool to register populators and Biomes but it doesn't really seem to be properly registering on the other thread since GeneratorRegisterTask is complaining about Biome::$biomes being uninitialized..

**Comment by dktapps** [comment]
like I said it's really not practical to do this. It would require making basically everything aware of the server and its async pool, which is not really a position I want to put the code in.

**Comment by TheShermanTanker** [comment]
Would an explicit pass be acceptable in this instance?

**Comment by dktapps** [comment]
I have no idea what you're talking about and I also have no idea how are you trying to implement this. the preferred solution to this problem is to use AsyncPool->addWorkerStartHook() to schedule an async task to do what's needed. But your generator class can do that anyway.

**Comment by dktapps** [comment]
I think this isn't really any different than registering custom blocks, which also can't be trivially copied to other threads. The status quo is, as I mentioned above, to use `AsyncPool->addWorkerStartHook()` to schedule an async task to register your custom stuff. Also, since 4.0 the `BiomeRegistry` API is public, so I think this issue can be closed.
